flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecMultiJoin.java

```diff
@@ -0,0 +1,388 @@
+
+import org.apache.calcite.rex.RexNode;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Stream {@link StreamExecNode} for N-way Joins. This node handles multi-way joins in streaming
+ * mode, supporting different join types and conditions for each input.
+ */
+@ExecNodeMetadata(
+        name = "stream-exec-multi-join",
+        version = 1,
+        producedTransformations = StreamExecMultiJoin.MULTI_JOIN_TRANSFORMATION,
+        minPlanVersion = FlinkVersion.v2_1,
+        minStateVersion = FlinkVersion.v2_1)
+public class StreamExecMultiJoin extends ExecNodeBase<RowData>
+        implements StreamExecNode<RowData>, SingleTransformationTranslator<RowData> {
+
+    public static final String MULTI_JOIN_TRANSFORMATION = "multi-join";
+    private static final String FIELD_NAME_JOIN_TYPES = "joinTypes";
+    private static final String FIELD_NAME_JOIN_CONDITIONS = "joinConditions";
+    private static final String FIELD_NAME_JOIN_ATTRIBUTE_MAP = "joinAttributeMap";
+    private static final String FIELD_NAME_INPUT_UPSERT_KEYS = "inputUpsertKeys";
+    private static final String FIELD_NAME_STATE_METADATA_LIST = "stateMetadataList";
+    private static final String FIELD_NAME_MULTI_JOIN_CONDITION = "multiJoinCondition";
+
+    @JsonProperty(FIELD_NAME_JOIN_TYPES)
+    private final List<JoinType> joinTypes;
+
+    @JsonProperty(FIELD_NAME_JOIN_CONDITIONS)
+    private final List<? extends @Nullable RexNode> joinConditions;
+
+    @SuppressWarnings({"unused", "FieldCanBeLocal"})
+    @JsonProperty(FIELD_NAME_MULTI_JOIN_CONDITION)
+    private final RexNode multiJoinCondition;
+
+    @JsonProperty(FIELD_NAME_JOIN_ATTRIBUTE_MAP)
+    private final Map<Integer, List<ConditionAttributeRef>> joinAttributeMap;
+
+    @JsonProperty(FIELD_NAME_INPUT_UPSERT_KEYS)
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    // List of upsert keys for each input, where each inner list corresponds to an input
+    // The reason it's a List<List<int[]>> is that SQL allows only one primary key but
+    // multiple upsert (unique) keys per input
+    private final List<List<int[]>> inputUpsertKeys;
+
+    @JsonProperty(FIELD_NAME_STATE_METADATA_LIST)
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private final List<StateMetadata> stateMetadataList;
+
+    public StreamExecMultiJoin(
+            final ReadableConfig tableConfig,
+            final List<JoinType> joinTypes,
+            final List<? extends @Nullable RexNode> joinConditions,
+            @Nullable final RexNode multiJoinCondition,
+            final Map<Integer, List<AttributeBasedJoinKeyExtractor.ConditionAttributeRef>>
```

Comment:
```suggestion
            final Map<Integer, List<ConditionAttributeRef>>
```
also below because you important the class already, which makes sense for readibility